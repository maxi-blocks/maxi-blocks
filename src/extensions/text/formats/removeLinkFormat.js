/**
 * WordPress dependencies
 */
import { removeFormat } from '@wordpress/rich-text';

/**
 * Internal dependencies
 */
import getFormatPosition from './getFormatPosition';
import setFormat from './setFormat';

/**
 * External dependencies
 */
import { isEqual } from 'lodash';
import getCustomFormatValue from './getCustomFormatValue';
import getFormattedString from './getFormattedString';

/**
 * Removes the link and custom formats
 *
 * @param {Object} [$0]             Optional named arguments.
 * @param {Object} [$0.formatValue] RichText format value
 * @param {Object} [$0.typography]  MaxiBlocks typography
 * @param {Object} [$0.isList]      Text Maxi block has list mode active
 * @returns {Object} Returns cleaned and formatted typography and content
 */
const removeLinkFormat = ({
	formatValue,
	typography,
	isList,
	textLevel,
	attributes,
	blockStyle,
	styleCard,
}) => {
	const { start, end } = formatValue;
	const formatLength = formatValue.formats.length;
	let newStart = start;
	let newEnd = end;
	const [linkPositionStart, linkPositionEnd] =
		getFormatPosition({
			formatValue,
			formatName: 'maxi-blocks/text-link',
			formatClassName: null,
			formatAttributes: attributes,
		}) || [];
	if (start === end)
		[newStart, newEnd] = [linkPositionStart, linkPositionEnd] || [
			0,
			formatLength,
		];

	const removedLinkFormatValue = removeFormat(
		{
			...formatValue,
			start: newStart,
			end: newEnd,
		},
		'maxi-blocks/text-link'
	);

	// Check if the removing is a partial part of the whole content
	const isWholeContentLink = isEqual(
		[linkPositionStart, linkPositionEnd],
		[0, formatLength]
	);
	const hasUnderline =
		getCustomFormatValue({
			formatValue: { ...formatValue, start: 0, end: formatLength },
			typography,
			prop: 'text-decoration',
			breakpoint: 'general',
			blockStyle,
			textLevel,
			styleCard,
		}) === 'underline';
	const isPartialOfWholeLink =
		isWholeContentLink &&
		hasUnderline &&
		newStart !== 0 &&
		newEnd !== formatLength;
	const isLastPartialOfWholeLink =
		isWholeContentLink &&
		hasUnderline &&
		newStart !== 0 &&
		newStart !== formatLength &&
		newEnd === formatLength;

	if (isPartialOfWholeLink || isLastPartialOfWholeLink) {
		// Remove whole content underline
		const cleanTypography = setFormat({
			disableCustomFormats: true,
			isList,
			typography,
			value: {
				'text-decoration': '',
			},
			textLevel,
		});
		// Set first part
		const { formatValue: firstPartFormatValue, ...firstPartTypography } =
			start !== 0
				? setFormat({
						formatValue: {
							...removedLinkFormatValue,
							start: 0,
							end: start,
						},
						isList,
						typography: cleanTypography,
						value: {
							'text-decoration': 'underline',
						},
						textLevel,
						returnFormatValue: true,
				  })
				: { formatValue: removedLinkFormatValue, ...cleanTypography };
		// Set second part
		const secondPartTypography =
			end !== formatLength
				? setFormat({
						formatValue: {
							...firstPartFormatValue,
							start: end,
							end: formatLength,
						},
						isList,
						typography: firstPartTypography,
						value: {
							'text-decoration': 'underline',
						},
						textLevel,
				  })
				: { ...firstPartTypography };

		return { ...secondPartTypography };
	}

	if (
		isWholeContentLink &&
		hasUnderline &&
		((newStart === 0 && newEnd === formatLength) || start === end)
	) {
		const removedLinkContent = getFormattedString({
			formatValue: removedLinkFormatValue,
			isList,
		});
		return {
			...setFormat({
				disableCustomFormats: true,
				isList,
				typography,
				value: {
					'text-decoration': '',
				},
				textLevel,
			}),
			content: removedLinkContent,
		};
	}

	return setFormat({
		formatValue: removedLinkFormatValue,
		isList,
		typography,
		value: {
			'text-decoration': '',
		},
		textLevel,
	});
};

export default removeLinkFormat;
