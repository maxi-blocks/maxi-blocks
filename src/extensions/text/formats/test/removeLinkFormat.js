import removeLinkFormat from '../removeLinkFormat';

jest.mock('@wordpress/data', () => ({
	dispatch: jest.fn(() => ({
		__unstableMarkLastChangeAsPersistent: jest.fn(),
	})),
	select: jest.fn(() => ({
		receiveMaxiSelectedStyleCard: jest.fn(() => ({ value: {} })),
	})),
}));
jest.mock('@extensions/styles', () => ({
	getBlockStyle: jest.fn(() => 'light'),
	getLastBreakpointAttribute: jest.fn(
		({ target, breakpoint, attributes }) =>
			attributes[`${target}-${breakpoint}`]
	),
}));
jest.mock('@extensions/style-cards', () => ({
	updateSCOnEditor: jest.fn(),
	getTypographyFromSC: jest.fn(() => ({})),
}));

const partialLinkFormatObject = {
	attributes: {
		url: 'https://test.com',
		opensInNewTab: false,
		title: '',
	},
	formatValue: {
		formats: Array(21).fill(
			[
				{
					type: 'maxi-blocks/text-custom',
					attributes: {
						className: 'maxi-text-block__custom-format--0',
					},
				},
			],
			6,
			12
		),
		replacements: Array(21),
		text: 'lorem ipsum dolor sit',
		start: 6,
		end: 12,
		activeFormats: [
			{
				type: 'maxi-blocks/text-link',
				attributes: {
					url: 'https://test.com',
					title: '',
				},
			},
			{
				type: 'maxi-blocks/text-custom',
				attributes: {
					className: 'maxi-text-block__custom-format--0',
				},
			},
		],
	},
	blockStyle: 'light',
	textLevel: 'p',
	styleCard: {
		light: {
			defaultStyleCard: {
				p: {},
			},
		},
	},
	typography: {
		customFormats: {
			'maxi-text-block__custom-format--0': {
				'text-decoration-general': 'underline',
			},
		},
	},
};
const wholeLinkFormatObject = {
	...partialLinkFormatObject,
	formatValue: {
		...partialLinkFormatObject.formatValue,
		activeFormats: [
			{
				type: 'maxi-blocks/text-link',
				attributes: {
					url: 'https://test.com',
					title: '',
				},
			},
		],
		formats: Array(21).fill([
			{
				type: 'maxi-blocks/text-link',
				attributes: {
					url: 'https://test.com',
					title: '',
				},
			},
		]),
		start: 1,
		end: 1,
	},
	typography: {
		'text-decoration-general': 'underline',
		customFormats: {},
	},
};

describe('removeLinkFormat', () => {
	let windowSpy;

	beforeEach(() => {
		windowSpy = jest.spyOn(window, 'window', 'get');
	});

	afterEach(() => {
		windowSpy.mockRestore();
	});

	it('Should remove custom formats when removing partial link', () => {
		const result = removeLinkFormat(partialLinkFormatObject);

		expect(result).toMatchSnapshot();
	});

	it('Should remove custom formats when removing whole link', () => {
		windowSpy.mockImplementation(() => ({
			getSelection: () => ({
				getRangeAt: () => ({
					startOffset: 1,
					endOffset: 1,
				}),
			}),
		}));

		const result = removeLinkFormat(wholeLinkFormatObject);

		expect(result).toMatchSnapshot();
	});

	it('Should remove custom formats when removing partial of whole link', () => {
		const result = removeLinkFormat({
			...wholeLinkFormatObject,
			formatValue: {
				...wholeLinkFormatObject.formatValue,
				start: 6,
				end: 12,
			},
		});

		expect(result).toMatchSnapshot();
	});
});
