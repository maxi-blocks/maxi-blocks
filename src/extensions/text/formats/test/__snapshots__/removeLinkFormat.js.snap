// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeLinkFormat Should remove custom formats when removing partial link 1`] = `
{
  "content": "lorem ipsum dolor sit",
  "custom-formats": {},
}
`;

exports[`removeLinkFormat Should remove custom formats when removing partial of whole link 1`] = `
{
  "content": "<maxi-blocks/text-link url="https://test.com" title=""><maxi-blocks/text-custom className="maxi-text-block__custom-format--1">lorem </maxi-blocks/text-custom></maxi-blocks/text-link>ipsum <maxi-blocks/text-link url="https://test.com" title=""><maxi-blocks/text-custom className="maxi-text-block__custom-format--1">dolor sit</maxi-blocks/text-custom></maxi-blocks/text-link>",
  "custom-formats": {
    "maxi-text-block__custom-format--1": {
      "text-decoration-general": "underline",
    },
  },
  "text-decoration-general": "",
}
`;

exports[`removeLinkFormat Should remove custom formats when removing whole link 1`] = `
{
  "content": "lorem ipsum dolor sit",
  "text-decoration-general": "",
}
`;
